src/core/atomic.d

```diff
@@ -866,6 +952,85 @@ else version (AsmX86_64)
         }
     }
 
+    shared(T) atomicExchange(MemoryOrder ms = MemoryOrder.seq,T,V)( shared(T)* here, V exchangeWith ) pure nothrow @nogc @safe
+        if ( !is(T == class) && !is(T U : U*) &&  __traits( compiles, { *here = exchangeWith; } ) )
+    {
+        return atomicExchangeImpl(here, exchangeWith);
+    }
+
+    shared(T) atomicExchange(MemoryOrder ms = MemoryOrder.seq,T,V)( shared(T)* here, shared(V) exchangeWith ) pure nothrow @nogc @safe
+        if ( is(T == class) && __traits( compiles, { *here = exchangeWith; } ) )
+    {
+        return atomicExchangeImpl(here, exchangeWith);
+    }
+
+    shared(T) atomicExchange(MemoryOrder ms = MemoryOrder.seq,T,V)( shared(T)* here, shared(V)* exchangeWith ) pure nothrow @nogc @safe
+        if ( is(T U : U*) && __traits( compiles, { *here = exchangeWith; } ) )
+    {
+        return atomicExchangeImpl(here, exchangeWith);
+    }
+
+    private shared(T) atomicExchangeImpl(T,V)( shared(T)* here, V exchangeWith ) pure nothrow @nogc @safe
+        in ( atomicPtrIsProperlyAligned( here ), "Argument `here` is not properly aligned" )
+    do
+    {
+        static if ( T.sizeof == byte.sizeof )
+        {
+            asm pure nothrow @nogc @trusted
+            {
+                mov AL, exchangeWith;
+                mov RCX, here;
+                xchg [RCX], AL;
+            }
+        }
+        else static if ( T.sizeof == short.sizeof )
+        {
+            asm pure nothrow @nogc @trusted
+            {
+                mov AX, exchangeWith;
+                mov RCX, here;
+                xchg [RCX], AX;
+            }
+        }
+        else static if ( T.sizeof == int.sizeof )
+        {
+            asm pure nothrow @nogc @trusted
+            {
+                mov EAX, exchangeWith;
+                mov RCX, here;
+                xchg [RCX], EAX;
+            }
+            static if ( __traits(isFloating, T) )
+            {
+                asm pure nothrow @nogc @trusted
+                {
+                    mov exchangeWith, EAX;
+                }
+                return exchangeWith;
+            }
+        }
+        else static if ( T.sizeof == long.sizeof )
+        {
+            asm pure nothrow @nogc @trusted
+            {
+                mov RAX, exchangeWith;
+                mov RCX, here;
+                xchg [RCX], RAX;
+            }
+            static if ( __traits(isFloating, T) )
+            {
+                asm pure nothrow @nogc @trusted
+                {
+                    mov exchangeWith, RAX;
+                }
+                return exchangeWith;
```

Comment:
@WalterBright I think there might be a codegen issue with asm in functions... this here assigns RAX to a local, which emits correct code, but then the return doesn't actually return the argument correctly.

I reduced the snippet:
```d
extern(C) float floop(float* r, float x)
{
  asm
  {
    mov EAX, x;
    mov RCX, r;
    xchg [RCX], EAX;
    mov x, EAX;
  }
  return x;
}
```
And compile with DMD: `dmd asmtest.d -c -m64` and emit's this:
```asm
0000000000000000 <floop>:
   0:   55                      push   %rbp
   1:   48 8b ec                mov    %rsp,%rbp
   4:   48 83 ec 10             sub    $0x10,%rsp
   8:   48 89 7d f0             mov    %rdi,-0x10(%rbp)
   c:   f3 0f 11 45 f8          movss  %xmm0,-0x8(%rbp)
  11:   8b 45 f8                mov    -0x8(%rbp),%eax
  14:   48 8b 4d f0             mov    -0x10(%rbp),%rcx
  18:   87 01                   xchg   %eax,(%rcx)
  1a:   89 45 f8                mov    %eax,-0x8(%rbp)
  1d:   c9                      leaveq
  1e:   c3                      retq
```
There's a line missing that should load the float back into xmm0 before returning... I suspect DMD is assuming that the asm block did not clobber `x`.

This is on linux... but when I build for windows, I don't see this issue.

Reply:
What's also strange is that I can't repro this issue on my local machine!
Is it possible there's a regression in a DMD more recent than the version I'm running locally?
Has anything changed recently that might affect this?

Reply:
`-m32` does work properly:
```asm
00000000 <floop>:
   0:   55                      push   %ebp
   1:   8b ec                   mov    %esp,%ebp
   3:   8b 45 0c                mov    0xc(%ebp),%eax
   6:   8b 4d 08                mov    0x8(%ebp),%ecx
   9:   87 01                   xchg   %eax,(%ecx)
   b:   89 45 0c                mov    %eax,0xc(%ebp)
   e:   d9 45 0c                flds   0xc(%ebp)  ; <- here it is!
  11:   5d                      pop    %ebp
  12:   c3                      ret
```
So, it's just x64... and only on master DMD, my local compiler works correctly.